### PR TITLE
Fix bug with proxies

### PR DESCRIPTION
**Pull request by mario**
Fixes issue https://github.com/nextcloud/server/issues/7805 (or at least it should).

cc @rullzer @nickvergessen and @MorrisJobke because this breaks things for me :(

Signed-off-by: Mario Danic <mario@lovelyhq.com>

### CODE REVIEW

**Comment by mario** [comment]
Targetting 13, but can be merged to master too.

File: core/Controller/ClientFlowLoginController.php [hunk 1]
```diff
@@ -315,7 +315,19 @@ public function generateAppPassword($stateToken,
 				$serverPostfix = substr($this->request->getRequestUri(), 0, strpos($this->request->getRequestUri(), '/login/flow'));
 			}
 
-			$serverPath = $this->request->getServerProtocol() . "://" . $this->request->getServerHost() . $serverPostfix;
+			$protocol = $this->request->getServerProtocol();
+
+			if ($protocol != "https") {
+				$xForwardedProto = $this->request->getHeader('HTTP_X_FORWARDED_PROTO');
+				$xForwardedSSL = $this->request->getHeader('HTTP_X_FORWARDED_SSL');
+				if (($xForwardedProto !== null && $xForwardedProto == "https") ||
```
Comment (rullzer) [inline core/Controller/ClientFlowLoginController.php]: use `===`

File: core/Controller/ClientFlowLoginController.php [hunk 2]
```diff
@@ -315,7 +315,19 @@ public function generateAppPassword($stateToken,
...
+					($xForwardedSSL !== null && $xForwardedSSL == "on")) {
```
Comment (rullzer) [inline core/Controller/ClientFlowLoginController.php]: use `===`

Comment (rullzer) [reply core/Controller/ClientFlowLoginController.php]: Also why first test if it is `null` ;)

**Review by rullzer — DISMISSED** [review]
can you also add tests for this?

**Comment by mario** [comment]
@rullzer done!

**Comment by mario** [comment]
Tests unlikely, but you can if I ask nicely?:)

**Comment by mario** [comment]
Of course, real testing also appreciated, since I can't test atm.

**Comment by nickvergessen** [comment]
Added the tests

**Comment by MorrisJobke** [comment]
> Targetting 13, but can be merged to master too.

Usually we do the one against master first and then backport to the stable branches. But let's do it the other way around here too.

**Review by rullzer — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]

**Pull request merged**